### ISSUE TITLE
graphviz_layout() selection

### ISSUE THREAD

**Issue by hagberg** [networkx/networkx#1596]
The graphviz_layout() algorithm selection code doesn't work right.
In the case below I have graphviz_layout() provided by nx_agraph and not one of the pydot packges.
The pydot package loader doesn't bomb on ImportError until run time.

```python
In [1]: import networkx as nx

In [2]: G = nx.path_graph(4)

In [3]: from networkx import graphviz_layout

In [4]: pos = graphviz_layout(G)
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-4-29bbca8a9cf2> in <module>()
----> 1 pos = graphviz_layout(G)

/home/aric/.local/lib/python2.7/site-packages/networkx-2.0.dev_20150613115110-py2.7.egg/networkx/drawing/nx_pydot.pyc in graphviz_layout(G, prog, root, **kwds)
    255     This is a wrapper for pydot_layout.
    256     """
--> 257     return pydot_layout(G=G,prog=prog,root=root,**kwds)
    258 
    259 

/home/aric/.local/lib/python2.7/site-packages/networkx-2.0.dev_20150613115110-py2.7.egg/networkx/drawing/nx_pydot.pyc in pydot_layout(G, prog, root, **kwds)
    269     >>> pos=nx.pydot_layout(G,prog='dot')
    270     """
--> 271     pydot = load_pydot()
    272 
    273     P=to_pydot(G)

/home/aric/.local/lib/python2.7/site-packages/networkx-2.0.dev_20150613115110-py2.7.egg/networkx/drawing/nx_pydot.pyc in load_pydot()
     45     else:
     46         msg = "pydot could not be loaded: http://code.google.com/p/pydot/"
---> 47         raise ImportError(msg)
     48 
     49     return module

ImportError: pydot could not be loaded: http://code.google.com/p/pydot/

In [5]: from networkx import nx_agraph

In [6]: pos = nx_agraph.graphviz_layout(G)
```

**Comment by chebee7i** [comment]
It looks like each graphviz_layout is independent of the others. Are these supposed to be chained, trying the next until something succeeds?

**Comment by hagberg** [comment]
Yes, that is how it used to work.
But the pydot updates there changed that.

It certainly wasn't optimal before either. It was hard to tell where you were getting graphviz_ functions from.

**Comment by chebee7i** [comment]
Something that I just noticed: The function signatures for `nx_pydot.graphviz_layout` and `nx_agraph.pygraphviz_layout` are not the same. We could probably make them match, but now I'm wondering if we should require the user to decide which `graphviz_layout` function should be called. In other words, we keep this as is?

**Comment by hagberg** [comment]
Good point, I didn't even notice the signatures.

If we allow the user to decide maybe then the imports should not be automatic.

Also the auto-select pydot code throws the import error at run time instead of import time.  Maybe that isn't good practice?

**Comment by chebee7i** [comment]
It looks like we do this for all the drawing functions now, even in `nx_agraph`. For example, `nx_agraph.to_agraph` has a function level import for `pygraphviz`.

If we don't want run time errors, we could make `drawing.__init__` try to import the various modules (pydot, pygraphviz) and if it fails, then the corresponding submodule is not imported. This should allow the submodules (`nx_agraph` etc) to assume that the needed package is available.

**Comment by hagberg** [comment]
I guess it was fcac1ee1d7e1f6ff30debee1139e5f80cdde93ae where this was changed.

**Comment by chebee7i** [comment]
Ah ok, that looks familiar. :) I'll make this a PR shortly.

**Comment by chebee7i** [comment]
Here's another thought...if we do this conditional import, then if pydot/pygraphviz is not available, the user will get an `AttributeError` when doing a call such as: `nx.to_agraph`. So this is still going to be a runtime error from the user perspective, esp if the call happens inside some function they have defined. 

I wonder if this might this be more confusing. E.g. the documentation suggests this function should be available, but it does not seem to be. Also, the error message no longer makes it clear why the exception arises---namely, that `pygraphviz` is unavailable.

Contrast this to the situation when the except is raised by the function when called. Here we explicitly tell them why it fails and that an optional package is needed.

**Comment by hagberg** [comment]
I'm open to other options.  Explicit errors (even imports) make the most sense to me.

**Comment by chebee7i** [comment]
Yeah it seems like not doing any imports in `nx.drawing` and requiring the user to explicitly import `nx_agraph` or `nx_pydot` could be more clear. This means these functions will no longer be available at the top level namespace. 

With 2.0 coming around, I wonder if this is the time to simply remove the drawing subpackage and make it an add-on with a required dependency of both (or maybe at least one of) `pygraphviz` and `pydot`.

**Comment by hagberg** [comment]
I've been advocating for removing the drawing package so fine with me.
It would be great to have a graph layout package too (potentially separate from that).

**Comment by chebee7i** [comment]
Ok let's do that. We can probably close this, and then link to it in a new issue that will turn `networkx.drawing` into `networkx.addons.drawing` and `networkx.addons.layouts`. I'm off to the Jurassic ...

```
                                _.-.
                               /  99\
                              (      `\
                              |\\ ,  ,|
                      __      | \\____/
                ,.--"`-.".   /   `---'
            _.-'          '-/      |
        _.-"   |   '-.             |_/_
  ,__.-'  _,.--\      \      ((    /-\
  ',_..--'      `\     \      \\_ /
                  `-,   )      |\'
                    |   |-.,,-" (
                    |   |   `\   `',_
                    )    \    \,(\(\-'
                jgs \     `-,_
                     \_(\-(\`-`
                        "  "
```

**Comment by OrkoHunter** [comment]
The Add-on project will hopefully cover this issue. Possibly after `networkx-lemon`.
@chebee7i Btw. Great movie! ;)

**Comment by hagberg** [comment]
Fixed with in #1931 with absolute imports
